File: 202504/grep/go/impl/grep_performance_measurement.go
package impl

import (
	"fmt"
	"io/ioutil"
	"strings"

	utils "study-session/utils/go"
)

// loadGrepTestData は入力ファイルと期待値ファイルを読み込む
func loadGrepTestData(fileDir string) (string, string, []string, error) {
	// 入力データの読み込み
	inputData, err := ioutil.ReadFile(strings.Join([]string{fileDir, "input.txt"}, "/"))
	if err != nil {
		return "", "", nil, fmt.Errorf("入力ファイルの読み込みに失敗しました: %v", err)
	}

	// 入力データのパース
	inputLines := strings.Split(string(inputData), "\n")
	if len(inputLines) < 2 {
		return "", "", nil, fmt.Errorf("入力データが不足しています")
	}

	filePath := strings.Join([]string{fileDir, strings.TrimSpace(inputLines[0])}, "/")
	pattern := strings.TrimSpace(inputLines[1])

	// 期待値の読み込み
	expectedData, err := ioutil.ReadFile(strings.Join([]string{fileDir, "expected.txt"}, "/"))
	if err != nil {
		return filePath, pattern, nil, fmt.Errorf("期待値ファイルの読み込みに失敗しました: %v", err)
	}

	// 期待値のパース
	var expectedOutput []string
	for _, line := range strings.Split(string(expectedData), "\n") {
		if line != "" {
			expectedOutput = append(expectedOutput, line)
		}
	}

	return filePath, pattern, expectedOutput, nil
}

// MeasureGrepPerformance はGrepの性能と正当性を計測する
func MeasureGrepPerformance(fileDir string, iterations int) map[string]interface{} {
	var err error
	filePath, pattern, expectedOutput, err := loadGrepTestData(fileDir)
	if err != nil {
		fmt.Println(err)
		return nil
	}

	grep := &GrepImplementation{}

	fmt.Printf("Grep実装のパフォーマンス計測と正当性検証:\n")
	fmt.Printf("ファイル: %s\n", filePath)
	fmt.Printf("検索パターン: %s\n", pattern)
	fmt.Printf("繰り返し回数: %d\n", iterations)

	var matchingLines []string

	// 処理時間とメモリ使用量を計測
	results := utils.MeasurePerformance("Grep", func() {
		for i := 0; i < iterations; i++ {
			matchingLines = grep.Search(filePath, pattern)
			if iterations == 1 {
				fmt.Printf("ヒット数: %d\n", len(matchingLines))
			}
		}
	})

	// 正当性検証
	valid := utils.VerifyResult("Grep", matchingLines, expectedOutput)
	results["valid"] = valid

	return results
}

File: 202504/grep/ruby/grep_performance_measurement.rb
require_relative 'grep_implementation'
require_relative '../../utils/ruby/performance_measurement'
require_relative '../../utils/ruby/validity_check'

# Grep実装のパフォーマンスと正当性を計測するクラス
class GrepPerformanceMeasurement
  def initialize(file_path, implementation = GrepImplementation.new)
    @file_path = file_path
    @implementation = implementation
  end
  
  # 入力ファイルと期待値ファイルを読み込む
  def load_test_data
    input_path = [@file_path, 'input.txt'].join(File::SEPARATOR)
    expected_path = [@file_path, 'expected.txt'].join(File::SEPARATOR)

    # 入力データの読み込み
    input_data = File.read(input_path).strip
    # 入力データをパースして、ファイルパスとパターンを取得
    file_name, pattern = input_data.split("\n", 2)
    pattern.strip! if pattern

    file_path = [@file_path, file_name].join(File::SEPARATOR)
    
    # 期待値の読み込み
    expected_output = []
    if File.exist?(expected_path)
      expected_output = File.readlines(expected_path).map(&:chomp)
    end
    
    return file_path, pattern, expected_output
  end
  
  # パフォーマンス計測と正当性検証を実行
  def run_measurement(iterations = 1)
    # テストデータが指定されていない場合は読み込む
    begin
      file_path, pattern, expected_output = load_test_data
    rescue => e
      puts "テストデータの読み込みに失敗しました: #{e.message}"
      return nil
    end
    
    puts "Grep実装のパフォーマンス計測と正当性検証:"
    puts "ファイル: #{file_path}"
    puts "検索パターン: #{pattern}"
    puts "繰り返し回数: #{iterations}"
    
    matching_lines = nil
    
    # 処理時間とメモリ使用量を計測
    results = PerformanceMeasurement.measure("Grep") do
      iterations.times do |i|
        matching_lines = @implementation.search(file_path, pattern)
        # 最後の反復結果を保持
        if i == iterations - 1
          puts "ヒット数: #{matching_lines.length}"
        end
      end
    end
    
    # 正当性検証
    valid = ValidityCheck.verify("Grep", matching_lines, expected_output)
    results[:valid] = valid
    
    results
  end
end

# GrepPerformanceMeasurementクラスを使用して、Grepのパフォーマンスと正当性を計測する
if __FILE__ == $0
  puts "=============================="
  puts "Grepパフォーマンス計測"
  puts "=============================="
  
  # Grep計測と検証
  puts "Grep実装のテスト"
  file_path = ARGV[0] || ''
  grep_measurement = GrepPerformanceMeasurement.new(file_path)
  grep_results = grep_measurement.run_measurement
  
  # 検証結果の要約
  puts "\n=============================="
  puts "テスト結果サマリー"
  puts "=============================="
  puts "Grep: #{grep_results && grep_results[:valid] ? '成功 ✓' : '失敗 ✗'}"
end
